### PR TITLE
Feature/76 read data from confluence into vector database

### PR DESCRIPTION
**Pull request by Kadi-7**
All functionalities have been implemented except for the embeddings.

### CODE REVIEW

File: QAChat/Data_Processing/confluence_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                #           f.write(bits)
+                pdf_file = io.BytesIO(r.content)
+                pdf_reader = PyPDF2.PdfReader(pdf_file)
+                num_pages = len(pdf_reader.pages)
+
+                for page_num in range(num_pages):
+                    page = pdf_reader.pages[page_num]
+                    text = page.extract_text()
+                    content += text
+                pdf_content.append(content)
```
Comment (emuguy1) [inline QAChat/Data_Processing/confluence_preprocessor.py:171]: Why do you have this. The PDF Reading is implemented by Felix and me and only needs to be used. You just need to give the pdf to the function and your good to go.

File: QAChat/Data_Processing/confluence_preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+CONFLUENCE_PASSWORD = os.getenv("CONFLUENCE_PASSWORD")
+
+# Get Supabase API credentials from environment variables
+SUPABASE_URL = os.getenv("SUPABASE_URL")
+SUPABASE_SERVICE_KEY = os.getenv("SUPABASE_SERVICE_KEY")
+
+
+class ConfluencePreprocessor:
+    spaces = []
+    pages = []
```
Comment (emuguy1) [inline QAChat/Data_Processing/confluence_preprocessor.py:28]: This class should inherit from DataPreprocessor. so the constructor should look something like:
`classConfluencePreprocessor(DataPreprocessor):`
and then it should implement the method:
`def load_preprocessed_data(self, before: datetime, after: datetime) -> List[DataInformation]:`
When you have this, you can use the document embedded to create the embeddings and save it into the supabase database. Therefore you just need to add ypir preprocessor to the function:
`def store_information_in_database(self, typ: DataSource):` line 43 in the same way the DummyPrepocessor was added and then you can call it something like in the main.py of the Data_Processing folder.

**Review by emuguy1 — CHANGES_REQUESTED** [review]
Great Progress. But I think we should work on this some more, before we merge this.

**Comment by Felix-012** [comment]
I think your test is failing because it installs a hashlib version that is not compatible with python3. But I think hashlib is part of the standard library anyway so you should not need it in the requirements afaik

**Comment by emuguy1** [comment]
Great improvments

**Comment by Felix-012** [comment]
the test is still failing because hashlib is still in requirements.txt btw

**Comment by Felix-012** [comment]
but looks good otherwise

**Review by emuguy1 — APPROVED** [review]
Great work. Just some minor adjustments

File: QAChat/Data_Processing/pdf_reader.py [hunk 3]
```diff
@@ -31,7 +41,7 @@ def read_pdf(pdf_object, before: datetime, after: datetime, data_src: DataSource
     if len(text.strip()) < 100:
         text = __ocr_pdf(pdf_object)
 
-    return transform_to_data_information_list(text, before, after, data_src)
+    #return transform_to_data_information_list(text, before, after, data_src)
```
Comment (emuguy1) [inline QAChat/Data_Processing/pdf_reader.py:44]:
```suggestion
    return transform_to_data_information_list(text, before, after, data_src)
```

If you really want to handle the Data yourself, just use your other method. But you don't have to use it, as described in the other comment.

File: QAChat/Data_Processing/pdf_reader.py [hunk 4]
```diff
@@ -23,6 +23,16 @@
...
 
+def get_text_from_pdf(pdf_object):
+    # First try to read the PDF with PDFMiner
+    text = __read_pdf_from_object(pdf_object)
+
+    # If the result is empty or too short, try OCR
+    if len(text.strip()) < 100:
+        text = __ocr_pdf(pdf_object)
+
+    return text
```
[the remark this answers is not in the document]
Comment (zenzeii) [reply QAChat/Data_Processing/pdf_reader.py]: It was just for testing. Its removed again in the latest version

File: QAChat/Data_Processing/pdf_reader.py (hunk 3 above)
Comment (zenzeii) [reply QAChat/Data_Processing/pdf_reader.py]: Overlooked this and inadvertently pushed it. But its now reverted.

**Comment by zenzeii** [comment]
The PDF now can be stored in Supabase as chunks but it takes little more time than usual.

**Comment by zenzeii** [comment]
Forgot to comment here there was a minor issue in token.env (already updated in our secure channel)

The URL for confluence have to end with **/wiki/**
So [https://testqaware.atlassian.net/wiki/](https://testqaware.atlassian.net/wiki/) instead of [https://testqaware.atlassian.net/](https://testqaware.atlassian.net/).

Otherwise the PDF download link wont work in the code.

**Review by Felix-012 — APPROVED** [review]

File: requirements.txt [hunk 5]
```diff
@@ -25,6 +25,11 @@ transformers~=4.28.1
 python-dateutil~=2.8.2
 nltk~=3.8.1
 rapidfuzz~=2.13.7
+atlassian-python-api
+bs4
+psutil
+requests
+PyPDF2
```
Comment (emuguy1) [inline requirements.txt:32]: One last little tiny thing. As far as I can tell the atlassian-python-api and psutil seem to be duplicate and PyPDF2 seems to be an old added import. But other than that. very happy. Great job. If you agree you can just click commit suggestion.
```suggestion
bs4
requests
```

**Review by emuguy1 — APPROVED** [review]

File: requirements.txt (hunk 5 above)
Comment (zenzeii) [reply requirements.txt:32]: Indeed, seems like not all have been pushed and thanks for suggesting :)

**Pull request merged**